File: multiCamCommon/swift/Sources/MultiCamCommon/UploadTypes.swift
import Foundation

/// Upload item status values
public enum UploadStatus: String, Codable, CaseIterable {
    /// Upload is queued and waiting
    case queued = "queued"

    /// Upload is currently in progress
    case uploading = "uploading"

    /// Upload completed successfully
    case completed = "completed"

    /// Upload failed (see error field)
    case failed = "failed"
}

/// Upload item with progress information
///
/// Represents a single file upload in the device's upload queue.
public struct UploadItem: Codable {
    /// Filename
    public let fileName: String

    /// Total file size in bytes
    public let fileSize: Int64

    /// Bytes uploaded so far
    public let bytesUploaded: Int64

    /// Upload progress percentage (0-100)
    public let uploadProgress: Double

    /// Current upload speed in bytes per second
    public let uploadSpeed: Int64

    /// Upload status (queued, uploading, completed, failed)
    public let status: String

    /// Presigned S3 URL for upload (only present when using presigned URL auth)
    public let uploadUrl: String?

    /// Error message if upload failed
    public let error: String?

    public init(
        fileName: String,
        fileSize: Int64,
        bytesUploaded: Int64,
        uploadProgress: Double,
        uploadSpeed: Int64,
        status: String,
        uploadUrl: String? = nil,
        error: String? = nil
    ) {
        self.fileName = fileName
        self.fileSize = fileSize
        self.bytesUploaded = bytesUploaded
        self.uploadProgress = uploadProgress
        self.uploadSpeed = uploadSpeed
        self.status = status
        self.uploadUrl = uploadUrl
        self.error = error
    }

    /// Get the status as an UploadStatus enum value
    public var uploadStatus: UploadStatus? {
        return UploadStatus(rawValue: status)
    }
}

File: multiCamCommon/java/src/main/java/com/multicam/common/UploadTypes.java
package com.multicam.common;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import java.util.List;
import java.util.ArrayList;

/**
 * Upload-related data types for MultiCam API.
 */
public class UploadTypes {
    private static final Gson gson = new GsonBuilder().create();

    /**
     * Upload item status values.
     */
    public enum UploadStatus {
        /** Upload is queued and waiting */
        QUEUED("queued"),

        /** Upload is currently in progress */
        UPLOADING("uploading"),

        /** Upload completed successfully */
        COMPLETED("completed"),

        /** Upload failed (see error field) */
        FAILED("failed");

        private final String value;

        UploadStatus(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }

        public static UploadStatus fromValue(String value) {
            for (UploadStatus status : values()) {
                if (status.value.equals(value)) {
                    return status;
                }
            }
            return null;
        }
    }

    /**
     * Upload item with progress information.
     * <p>
     * Represents a single file upload in the device's upload queue.
     */
    public static class UploadItem {
        /** Filename */
        public String fileName;

        /** Total file size in bytes */
        public long fileSize;

        /** Bytes uploaded so far */
        public long bytesUploaded;

        /** Upload progress percentage (0-100) */
        public double uploadProgress;

        /** Current upload speed in bytes per second */
        public long uploadSpeed;

        /** Upload status (queued, uploading, completed, failed) */
        public String status;

        /** Presigned S3 URL for upload (only present when using presigned URL auth) */
        public String uploadUrl;

        /** Error message if upload failed */
        public String error;

        public UploadItem() {
        }

        public UploadItem(String fileName, long fileSize, long bytesUploaded,
                         double uploadProgress, long uploadSpeed, String status,
                         String uploadUrl, String error) {
            this.fileName = fileName;
            this.fileSize = fileSize;
            this.bytesUploaded = bytesUploaded;
            this.uploadProgress = uploadProgress;
            this.uploadSpeed = uploadSpeed;
            this.status = status;
            this.uploadUrl = uploadUrl;
            this.error = error;
        }

        /**
         * Get the status as an UploadStatus enum value.
         *
         * @return UploadStatus enum value, or null if unknown
         */
        public UploadStatus getUploadStatus() {
            return UploadStatus.fromValue(status);
        }
    }
}
